include/fi_tagged_rx.h

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (c) 2015-2016 Intel Corporation, Inc.  All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef _FI_TAGGED_RX_H_
+#define _FI_TAGGED_RX_H_
+
+#include <fi.h>
+#include <fi_mem.h>
+#include <rdma/fi_tagged.h>
+
+/*
+ * Tagged recv list template: provides a set of utility api's for
+ * software-level tag matching, managing recv entries and unexpected
+ * messages.
+ * 
+ * Caller must provider an entrytype of the following
+ * structure:
+ *
+ * struct <entry_type> {
+ *         <mandatory fields>
+ *         struct dlist_entry entry;
+ *         struct fi_msg_tagged tmsg;
+ *         int fs_entry;
+ *
+ *         <optional: iovec's and other field(s)>
+ *         struct iovec iov[MAX_IOV_LIMIT];
+ *         void *desc[MAX_IOV_LIMIT];
```

Comment:
Actually, can't we define those fields as part of another structure that embeds the user's structure?  Then the developer doesn't need to deal with them at all.